File: 3.x.x/crates-utopia/emails/src/canonicals/yahoo.rs
use crate::canonicals::provider::{
    remove_hyphen_subaddress, require_non_empty_local, to_lower_case, Canonical, Provider,
};
use crate::EmailError;

/// Yahoo canonicalization (PHP `Yahoo`).
#[derive(Debug, Clone, Copy, Default)]
pub struct Yahoo;

const SUPPORTED_DOMAINS: &[&str] = &[
    "yahoo.com",
    "yahoo.co.uk",
    "yahoo.ca",
    "yahoo.de",
    "yahoo.fr",
    "yahoo.in",
    "yahoo.it",
    "ymail.com",
    "rocketmail.com",
];
const CANONICAL_DOMAIN: &str = "yahoo.com";

impl Provider for Yahoo {
    fn supports(&self, domain: &str) -> bool {
        SUPPORTED_DOMAINS.contains(&domain)
    }

    fn get_canonical(&self, local: &str, _domain: &str) -> Result<Canonical, EmailError> {
        let normalized = to_lower_case(local);
        let normalized = remove_hyphen_subaddress(&normalized);
        let normalized = require_non_empty_local(normalized)?;
        Ok(Canonical {
            local: normalized,
            domain: CANONICAL_DOMAIN.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        CANONICAL_DOMAIN
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        SUPPORTED_DOMAINS
    }
}

File: packages/emails/src/Emails/Canonicals/Providers/Yahoo.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * Yahoo
 *
 * Handles Yahoo email normalization based on validator.js rules
 * - Removes hyphen-based subaddress (everything after last -)
 * - Preserves dots in local part
 * - Normalizes to yahoo.com domain
 */
class Yahoo extends Provider
{
    private const SUPPORTED_DOMAINS = [
        'yahoo.com', 'yahoo.co.uk', 'yahoo.ca', 'yahoo.de', 'yahoo.fr', 'yahoo.in', 'yahoo.it',
        'ymail.com', 'rocketmail.com',
    ];

    private const CANONICAL_DOMAIN = 'yahoo.com';

    public function supports(string $domain): bool
    {
        return in_array($domain, self::SUPPORTED_DOMAINS, true);
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        // Remove hyphen-based subaddress (everything after last -)
        $normalizedLocal = $this->removeHyphenSubaddress($normalizedLocal);

        // Ensure local part is not empty after normalization
        if (empty($normalizedLocal)) {
            throw new \InvalidArgumentException('Email local part cannot be empty after normalization');
        }

        return [
            'local' => $normalizedLocal,
            'domain' => self::CANONICAL_DOMAIN,
        ];
    }

    public function getCanonicalDomain(): string
    {
        return self::CANONICAL_DOMAIN;
    }

    public function getSupportedDomains(): array
    {
        return self::SUPPORTED_DOMAINS;
    }
}
